fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java

```diff
@@ -3087,8 +3091,123 @@ private void setOperationContextProperties(FHIROperationContext operationContext
     }
 
     @Override
-    public int doReindex(FHIROperationContext operationContext, OperationOutcome.Builder operationOutcomeResult, Instant tstamp, List<Long> indexIds,
-        String resourceLogicalId) throws Exception {
+    public int doReindex(FHIROperationContext operationContext, OperationOutcome.Builder operationOutcomeResult, Instant tstamp,
+            List<Long> indexIds, String resourceLogicalId) throws Exception {
+        int result = 0;
+        // Since the try logic is slightly different in the code paths, we want to dispatch to separate methods to simplify the logic.
+        if (indexIds == null) {
+            result = doReindexSingle(operationOutcomeResult, tstamp, resourceLogicalId);
+        } else {
+            result = doReindexList(operationOutcomeResult, tstamp, indexIds);
+        }
+        return result;
+    }
+
+    /**
+     * encapsulates the logic to process a list with graduated backoff through the full list of indexIds
+     *
+     * @param operationOutcomeResult
+     * @param tstamp
+     * @param indexIds
+     * @return
+     * @throws Exception
+     */
+    public int doReindexList(OperationOutcome.Builder operationOutcomeResult, Instant tstamp, List<Long> indexIds) throws Exception {
+        // If the indexIds are empty, assume zero.
+        if (indexIds.isEmpty()) {
+            return 0;
+        }
+
+        /*
+         * How the backoff works...
+         * indexIds[1,2,3,4,5,6,7,8,9,10]
+         *
+         * Pass 1 left=0, right=10, max=10
+         * Result: Deadlock
+         *
+         * Pass2 left=0, right=1, max=10
+         * Move over by 1
+         * Result: Pass
+         *
+         * Pass3 left=1, right=2, max=10
+         * Move over by 1
+         * Result: Pass
+         *
+         * ... all the way up to 10
+         *
+         * Return total count back to caller.
+         *
+         * @implNote tried divide and conquer and it caused it to try large pass fail, small pass succeed, and therefore chose a small linear pass.
+         */
+
+        // Maximum number to retry
+        final int TX_ATTEMPTS = 5;
+
+        int result = 0;
+        int max = indexIds.size();
+
+        int left = 0;
+        int right = max;
+
+        int window = 0;
+        while (left < max) {
```

Comment:
This could end up trying 5 attempts for each item in the list, which probably risks a client timeout if there's something catastrophic going on. I think you can do the left < max and attempt limit in one loop. This puts an upper bound on the amount of time we're likely to spend here.

Reply:
OK - we can do that, I'll raise it to 10 attempts then.

Reply:
refactored so it's one loop up to ten attempts... next commit.